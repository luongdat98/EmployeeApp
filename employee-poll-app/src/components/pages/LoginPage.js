import { connect } from "react-redux";
import { useState } from "react";
import { loginUser } from "../../actions/authActions";
import { Navigate } from "react-router-dom";
import "bootstrap/dist/css/bootstrap.min.css";
import "../../styles/LoginPage.css";

const LoginPage = (props)=>{
  const {isAuthed} = props;

  const [username, setUsername] = useState("sarahedo");
  const [password, setPassword] = useState("password123");
  const [status, setStatus] = useState(true);
  
  const changeUsername = (e) => {
    setUsername(e.target.value);
  };

  const changePassword = (e) => {
    setPassword(e.target.value);
  };

  const clickLogin = (e) =>{
    e.preventDefault();
    props.dispatch(loginUser(username, password));
    setUsername("");
    setPassword("");
    if(isAuthed){
        setStatus(true)
    }
    else{
      setStatus(false)
    }
  };

  if (isAuthed) {
    return <Navigate to={`/`}/>;
  }

   return (
      <div className="form-signin">
        <form>
          <h1 className="h3 mb-3 fw-normal">Please sign in</h1>
          <div className="form-floating">
            <input type="text" onChange={changeUsername} className="form-control" data-testid="usernam-testing" id="floatingInput" placeholder="Username" defaultValue="sarahedo"/>
            <label for="floatingInput">Username</label>
          </div>
          <div className="form-floating">
            <input type="password" onChange={changePassword} class="form-control" data-testid="password-testing" id="floatingPassword" placeholder="Password" defaultValue="password123"/>
            <label for="floatingPassword">Password</label>
          </div>
          {!status && !username && <div className="error" data-testid="error-testing">The authentication was not successful!</div>}
          <button onClick={clickLogin} data-testid="button-testing" className="w-100 btn btn-lg btn-primary" type="submit">Sign in</button>
        </form>
      </div>
    );
}
const mapStateToProps =({authedUser}) =>({
   isAuthed : !(authedUser === null)
 }
)
export default connect(mapStateToProps)(LoginPage);