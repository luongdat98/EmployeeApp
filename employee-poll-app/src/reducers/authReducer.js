import { SET_AUTHERED_USER, LOGGED_OUT_USER } from "../actions/authActions";

export default function authedUser(state= null, action) {
  switch (action.type) {
    case SET_AUTHERED_USER:
      return action.id
    case LOGGED_OUT_USER:
      return null;
    default:
      return state;
  }
}