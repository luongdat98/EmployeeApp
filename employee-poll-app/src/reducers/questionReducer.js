import {
  RETRIEVE_QUESTIONS,
  ADD_ANSWER_QUESTION,
  ADD_QUESTION,
} from "../actions/questionActions";

export function questions(state = {}, action) {
  switch (action.type) {
    case RETRIEVE_QUESTIONS:
      return {
        ...state,
        ...action.questions,
      };
    case ADD_QUESTION:
      return {
        ...state,
        [action.qid]: action.question,
      };
    case ADD_ANSWER_QUESTION:
      return {
        ...state,
        [action.qid]: {
          ...state[action.qid],
          [action.answer]: {
            ...state[action.qid][action.answer],
            votes: state[action.qid][action.answer].votes.concat(
              action.authedUser
            ),
          },
        },
      };
    default:
      return state;
  }
}
