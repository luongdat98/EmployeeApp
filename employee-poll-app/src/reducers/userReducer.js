import {
  RETRIEVE_USERS,
  ADD_ANSWER_USER,
  ADD_QUESTION_USER,
} from "../actions/userActions";

export function users(state = {}, action) {
  switch (action.type) {
    case RETRIEVE_USERS:
      return {
        ...state,
        ...action.users,
      };
    case ADD_QUESTION_USER:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          questions: state[action.authedUser].questions.concat(action.qid),
        },
      };
    case ADD_ANSWER_USER:
      return {
        ...state,
        [action.authedUser]: {
          ...state[action.authedUser],
          answers: {
            ...state[action.authedUser].answers,
            [action.qid]: action.answer,
          },
        },
      };
    default:
      return state;
  }
}
