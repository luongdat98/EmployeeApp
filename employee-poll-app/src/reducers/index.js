import { combineReducers } from "redux";
import authedUser from "./authReducer";
import {users} from "./userReducer";
import {questions} from "./questionReducer";

export default combineReducers({
    authedUser,
    users,
    questions
});