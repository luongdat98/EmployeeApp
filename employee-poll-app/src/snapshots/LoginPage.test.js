import * as React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import "@testing-library/jest-dom";
import { legacy_createStore as createStore } from "redux";
import reducers from '../reducers';
import applyMiddleware from '../middleware';
import LoginPage from "../components/pages/LoginPage";

const store = createStore(reducers, applyMiddleware);

describe("Testing Login Page Successfully", ()=>{
  test("Expect matched snapshot", async ()=>{
    const login = render(
        <MemoryRouter>
            <Provider store={store}>
              <LoginPage/>
            </Provider>
        </MemoryRouter>
    );
    expect(login).toMatchSnapshot();
  });
});

describe("Testing Login Page Fail", () => {
  test("Return error message when login fail", async () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <LoginPage />
        </Provider>
      </MemoryRouter>
    );
    
    const username = screen.getByTestId("usernam-testing");
    const password = screen.getByTestId("password-testing");
    const button = screen.getByTestId("button-testing");

    fireEvent.change(username, { target: { value: "sarahedo" } });
    fireEvent.change(password, { target: { value: "password123" } });
    fireEvent.click(button);

    const error = screen.getByTestId("error-testing");
    expect(error).toBeInTheDocument();
  });
});