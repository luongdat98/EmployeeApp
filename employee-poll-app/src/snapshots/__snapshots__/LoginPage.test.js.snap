// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login Page Successfully Expect matched snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-signin"
      >
        <form>
          <h1
            class="h3 mb-3 fw-normal"
          >
            Please sign in
          </h1>
          <div
            class="form-floating"
          >
            <input
              class="form-control"
              data-testid="usernam-testing"
              id="floatingInput"
              placeholder="Username"
              type="text"
              value="sarahedo"
            />
            <label
              for="floatingInput"
            >
              Username
            </label>
          </div>
          <div
            class="form-floating"
          >
            <input
              class="form-control"
              data-testid="password-testing"
              id="floatingPassword"
              placeholder="Password"
              type="password"
              value="password123"
            />
            <label
              for="floatingPassword"
            >
              Password
            </label>
          </div>
          <button
            class="w-100 btn btn-lg btn-primary"
            data-testid="button-testing"
            type="submit"
          >
            Sign in
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-signin"
    >
      <form>
        <h1
          class="h3 mb-3 fw-normal"
        >
          Please sign in
        </h1>
        <div
          class="form-floating"
        >
          <input
            class="form-control"
            data-testid="usernam-testing"
            id="floatingInput"
            placeholder="Username"
            type="text"
            value="sarahedo"
          />
          <label
            for="floatingInput"
          >
            Username
          </label>
        </div>
        <div
          class="form-floating"
        >
          <input
            class="form-control"
            data-testid="password-testing"
            id="floatingPassword"
            placeholder="Password"
            type="password"
            value="password123"
          />
          <label
            for="floatingPassword"
          >
            Password
          </label>
        </div>
        <button
          class="w-100 btn btn-lg btn-primary"
          data-testid="button-testing"
          type="submit"
        >
          Sign in
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
